src/highlight/highlightUtil.ts

```diff
@@ -58,3 +58,56 @@ export const centerHighlight = (shapes: Shape[]): Position => {
 export function isHighlight(annotation: Annotation): annotation is AnnotationHighlight {
     return annotation?.target?.type === Type.highlight;
 }
+
+export const dedupeRects = (rects: Shape[], threshold = 0.1): Shape[] => {
+    const dedupedRects: Shape[] = [];
+
+    rects.forEach(curr => {
+        const prev = dedupedRects.pop();
+        // empty list, push current
+        if (!prev) {
+            dedupedRects.push(curr);
+            // The return in forEach callback is acting like a continue in for loop
+            return;
+        }
+
+        const heightThreshold = Math.max(Math.round(prev.height * threshold), Math.round(curr.height * threshold), 1);
+
+        if (prev.x === curr.x && prev.width === curr.width && Math.abs(prev.y - curr.y) <= heightThreshold) {
+            // the same rect, push the larger one
+            dedupedRects.push(prev.height > curr.height ? prev : curr);
+        } else {
+            // different rects, push both
+            dedupedRects.push(prev);
+            dedupedRects.push(curr);
+        }
+    });
+
+    return dedupedRects;
+};
+
+export const combineRectsByRow = (rects: Shape[]): Shape[] => {
+    const dedupedRects = dedupeRects(rects);
+
+    const rows = dedupedRects.reduce((groups, rect) => {
+        const { y } = rect;
+        const roundedY = Math.round(y);
+        groups[roundedY] = groups[roundedY] || [];
+        groups[roundedY].push(rect);
+        return groups;
+    }, {} as Record<number, Shape[]>);
```

Comment:
Would we be able to get rid of this cast if we defined `rows` as `Record<number, Shape[]>`?

Reply:
Tried, doesn't work. It seems like TS is not able to inter accumulator's type from result's type, unfortunately.